rcl/src/rcl/publisher.c

```diff
@@ -217,35 +212,49 @@ rcl_publish(const rcl_publisher_t * publisher, const void * ros_message)
 const char *
 rcl_publisher_get_topic_name(const rcl_publisher_t * publisher)
 {
-  const rcl_publisher_options_t * options = rcl_publisher_get_options(publisher);
-  if (!options) {
+  if (!rcl_publisher_is_valid(publisher)) {
     return NULL;
   }
-  RCL_CHECK_FOR_NULL_WITH_MSG(
-    publisher->impl, "publisher is invalid", return NULL, options->allocator);
-  RCL_CHECK_FOR_NULL_WITH_MSG(
-    publisher->impl->rmw_handle, "publisher is invalid", return NULL, options->allocator);
   return publisher->impl->rmw_handle->topic_name;
 }
 
+/* *INDENT-OFF* */
+#define _publisher_get_options(pub) &pub->impl->options
+/* *INDENT-ON* */
+
 const rcl_publisher_options_t *
 rcl_publisher_get_options(const rcl_publisher_t * publisher)
 {
-  RCL_CHECK_ARGUMENT_FOR_NULL(publisher, NULL, rcl_get_default_allocator());
-  RCL_CHECK_FOR_NULL_WITH_MSG(
-    publisher->impl, "publisher is invalid", return NULL, rcl_get_default_allocator());
-  return &publisher->impl->options;
+  if (!rcl_publisher_is_valid(publisher)) {
+    return NULL;
+  }
+  return _publisher_get_options(publisher);
 }
 
 rmw_publisher_t *
 rcl_publisher_get_rmw_handle(const rcl_publisher_t * publisher)
 {
-  RCL_CHECK_ARGUMENT_FOR_NULL(publisher, NULL, rcl_get_default_allocator());
-  RCL_CHECK_FOR_NULL_WITH_MSG(
-    publisher->impl, "publisher is invalid", return NULL, rcl_get_default_allocator());
+  if (!rcl_publisher_is_valid(publisher)) {
+    return NULL;
+  }
   return publisher->impl->rmw_handle;
 }
 
+bool
+rcl_publisher_is_valid(const rcl_publisher_t * publisher)
+{
+  const rcl_publisher_options_t * options;
+  RCL_CHECK_ARGUMENT_FOR_NULL(publisher, false, rcl_get_default_allocator());
+  options = _publisher_get_options(publisher);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    options, "publisher is invalid", return false, rcl_get_default_allocator());
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    publisher->impl, "publisher is invalid", return false, options->allocator);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    publisher->impl->rmw_handle, "publisher is invalid", return false, options->allocator);
```

Comment:
It would be better if the different checks use a different error message. Otherwise the user won't be able to distinguish the error case without additional debugging.